clippy_lints/src/recursive_format_impl.rs

```diff
@@ -0,0 +1,269 @@
+        if let ExprKind::MethodCall(path, _, [ref self_arg, ..], _) = expr.kind;
+        // Is the method to_string() ?
+        if path.ident.name == sym!(to_string);
+        // Is the method a part of the ToString trait? (i.e. not to_string() implemented
+        // separately)
+        if let Some(expr_def_id) = cx.typeck_results().type_dependent_def_id(expr.hir_id);
+        if is_diag_trait_item(cx, expr_def_id, sym::ToString);
+        // Is the method is called on self
+        if path_to_local_id(self_arg, self_hir_id);
+        then {
+            span_lint(
+                cx,
+                RECURSIVE_FORMAT_IMPL,
+                expr.span,
+                "using `to_string` in `fmt::Display` implementation might lead to infinite recursion",
+            );
+        }
+    }
+}
+
+fn check_self_in_format_args(cx: &LateContext<'_>, expr: &Expr<'_>, self_hir_id: HirId, impl_trait: ImplTrait) {
+    // Check each arg in format calls - do we ever use Display on self (directly or via deref)?
+    if_chain! {
+            if let Some(format_args) = FormatArgsExpn::parse(cx, expr);
+            let expr_expn_data = expr.span.ctxt().outer_expn_data();
+            let outermost_expn_data = outermost_expn_data(expr_expn_data);
+            if let Some(macro_def_id) = outermost_expn_data.macro_def_id;
+            if FORMAT_MACRO_PATHS
+                .iter()
+                .any(|path| match_def_path(cx, macro_def_id, path))
+                || FORMAT_MACRO_DIAG_ITEMS
+                    .iter()
+                    .any(|diag_item| cx.tcx.is_diagnostic_item(*diag_item, macro_def_id));
+            if let ExpnKind::Macro(_, _name) = outermost_expn_data.kind;
+            if let Some(args) = format_args.args();
+            then {
+                for (_i, arg) in args.iter().enumerate() {
+                    match impl_trait {
+                    // In Display, we only care about Display (it is okay to use Debug)
+                        ImplTrait::Display => {
+                    if arg.format_trait != sym::Display {
+                        continue;
+                    }},
+                    // In Debug, we only care about Debug (it is okay to use Display and ToString)
+                    ImplTrait::Debug => {
+                    if arg.format_trait != sym::Debug {
+                        continue;
+                    }},
+
+                    };
+                    check_format_arg_self(cx, expr, self_hir_id, arg, impl_trait);
+                }
+            }
+    }
+}
+
+fn check_format_arg_self(
+    cx: &LateContext<'_>,
+    expr: &Expr<'_>,
+    self_hir_id: HirId,
+    arg: &FormatArgsArg<'_>,
+    impl_trait: ImplTrait,
+) {
+    // Handle multiple dereferencing of references e.g. &&self
+    // Handle single dereference of &self -> self that is equivalent (i.e. via *self in fmt() impl)
+    // Since the argument to fmt is itself a reference: &self
+    let reference = single_deref(deref_expr(arg.value));
```

Comment:
You should handle *any* number of borrow/derefs (loop and `match e.kind`). Just make sure that the expression behind the deref operator is a reference.

Reply:
But you can have a case like:
```rust
struct J4 {}

impl std::ops::Deref for J4 {
    type Target = str;

    fn deref(&self) -> &Self::Target {
        "test"
    }
}

impl std::fmt::Display for J4 {
    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
        write!(f, "{}", &&**&&*self)
    }
}
```

I'm not sure how to catch this too since we have two derefs next to each other, so it's not as simple as only handling if it the next part is a reference :confused: 

This is unlikely to come up in practice though, and would trigger the immediate deref lint anyway.

Reply:
For one of those deref operators, the expression behind it is not a reference. `expr_ty(e).is_ref()`.

Reply:
Fixed, so the only remaining issue is `root_macro_call` vs `root_macro_call_first_node` - which expr should be passed if the latter is used? (passing the macro expr resulted in no matches).

Reply:
Oh I'm realizing this late but we also have an existing util for this - `peel_ref_operators`.

(I responded in the other thread about `root_macro_call`)

Reply:
Thanks, it's perfect for this case.